imgee/tasks.py

```diff
@@ -4,32 +4,58 @@
 
 class TaskRegistry(object):
     def __init__(self, name='default', connection=None):
-        self.connection = redis.from_url(connection) if connection else None
+        self.connection = connection or self.set_connection_from_url()
         self.name = name
-        self.key = 'imgee:registry:%s' % name
+        self.key_prefix = 'imgee:registry:%s' % name
 
-    def set_connection(self, connection=None):
-        connection = connection or app.config.get('REDIS_URL')
-        self.connection = redis.from_url(connection)
+    def set_connection_from_url(self, url=None):
+        url = url or app.config.get('REDIS_URL')
+        self.connection = redis.from_url(url)
+        return self.connection
 
-    def add(self, taskid):
-        self.connection.sadd(self.key, taskid)
-
-    def remove(self, taskid):
-        self.connection.srem(self.key, taskid)
-
-    def remove_all(self):
-        for k in self.get_all_keys():
-            self.remove(k)
+    def _make_key(self, taskid):
+        return "{key_prefix}:{taskid}".format(key_prefix=self.key_prefix, taskid=taskid)
 
     def __contains__(self, taskid):
-        return self.connection.sismember(self.key, taskid)
+        return len(self.connection.keys(self._make_key(taskid))) > 0
 
-    def keys_starting_with(self, query):
-        return filter(lambda k: k.startswith(query), self.connection.smembers(self.key))
+    def add(self, taskid, expire=60):
+        self.connection.set(self._make_key(taskid), taskid)
+        # setting TTL of 60 seconds for the key
+        # if the file doesn't get processed within 60 seconds,
+        # it'll be removed from the eregistry
+        self.connection.expire(self._make_key(taskid), expire)
 
     def search(self, query):
-        return filter(lambda k: str(query) in k, self.connection.smembers(self.key))
+        # >> KEYS imgee:registry:default:*query*
+        return self.connection.keys(self._make_key("*{}*".format(query)))
 
     def get_all_keys(self):
-        return list(self.connection.smembers(self.key))
+        # >> KEYS imgee:registry:default:*
+        return self.connection.keys(self._make_key("*"))
+
+    def keys_starting_with(self, query):
+        # >> KEYS imgee:registry:default:query*
+        return self.connection.keys(self._make_key("{}*".format(query)))
+
+    def remove(self, taskid):
+        # remove a key with the taskid
+        self.remove_by_key(self._make_key(taskid))
+
+    def remove_by_key(self, key):
+        # remove a redis key directly
+        self.connection.delete(key)
+
+    def remove_keys_matching(self, query):
+        keys = self.search(query)
+        for k in keys:
+            self.remove_by_key(k)
```

Comment:
Use a Redis pipe and push all commands at once, unless there's a way to do a wildcard remove directly in Redis.